lib/outdated.js

```diff
@@ -1,437 +1,273 @@
-
-    shouldUpdate(args, dep, name, has, required, depth, path, opts, cb, types[name])
-  }, cb)
-}
-
-function shouldUpdate (args, tree, dep, has, req, depth, pkgpath, opts, cb, type) {
-  // look up the most recent version.
-  // if that's what we already have, or if it's not on the args list,
-  // then dive into it.  Otherwise, cb() with the data.
-
-  // { version: , from: }
-  var curr = has[dep]
-
-  function skip (er) {
-    // show user that no viable version can be found
-    if (er) return cb(er)
-    outdated_(args,
-      pkgpath,
-      tree,
-      has,
-      depth + 1,
-      opts,
-      cb)
+  // packument fetching memoizing
+  const packuments = new Map()
+  async function getPackument (spec) {
+    if (packuments.has(spec)) {
+      return packuments.get(spec)
+    }
+    const packument = await pacote.packument(spec, {
+      fullMetadata: npm.flatOptions.long,
+      preferOnline: true
+    })
+    packuments.set(spec, packument)
+    return packument
   }
 
-  if (args.length && args.indexOf(dep) === -1) return skip()
+  async function getOutdatedInfo (name, edge) {
+    const spec = npa(name)
+    const node = edge.to
+    const {
+      path,
+      location,
+      package: {
+        version: current
+      }
+    } = node || { package: {} }
 
-  if (tree.isLink && req == null) return skip()
+    const type = edge.optional ? 'optionalDependencies'
+      : edge.peer ? 'peerDependencies'
+      : edge.dev ? 'devDependencies'
+      : 'dependencies'
 
-  if (req == null || req === '') req = '*'
+    // deps different from prod not currently
+    // on disk are not included in the output
+    if (edge.error === 'MISSING' && type !== 'dependencies') return
+    
+    try {
+      const packument = await getPackument(spec)
+      const expected = edge.spec
+      const wanted = pickManifest(packument, expected, npm.flatOptions)
+      const latest = pickManifest(packument, '*', npm.flatOptions)
+
+      if (
+        !current ||
+        current !== wanted.version ||
+        wanted.version !== latest.version
+      ) {
+        list.push({
+          name,
+          path,
+          type,
+          current,
+          location,
+          wanted: wanted.version,
+          latest: latest.version,
+          dependent: edge.from.name,
+          homepage: packument.homepage
+        })
+      }
+    } catch (err) {
+      // silently catch and ignore ETARGET, E403 &
+      // E404 errors, deps are just skipped
+      if (!(
+        err.code === 'ETARGET' ||
+        err.code === 'E403' ||
+        err.code === 'E404')
+      ) {
+        throw (err)
+      }
+    }
 
-  var parsed = npa.resolve(dep, req)
-  if (parsed.type === 'directory') {
-    if (tree.isLink) {
-      return skip()
-    } else {
-      return doIt('linked', 'linked')
+  }
+  
+  const p = []
+  if (deps.length !== 0) {
+    // specific deps
+    for (let i = 0; i < deps.length; i++) {
+      const nodes = tree.inventory.query('name', deps[i])
```

Comment:
I was initially a little curious as to why you had that in/out switch in `getEdges`, but it makes sense seeing this.  This is a lot more efficient than looping over every node in the tree!